mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -726,6 +887,76 @@ def set_tag(self, run_id, tag):
             self._check_run_is_active(run)
             session.merge(SqlTag(run_uuid=run_id, key=tag.key, value=tag.value))
 
+    def _set_tags(self, run_id, tags, attempt=None):
+        """
+        Set multiple tags on a run
+
+        :param run_id: String ID of the run
+        :param tags: List of RunTag instances to log
+        """
+        if not tags:
+            return
+
+        # use SessionManager instead of ManagedSessionManager to set
+        # isolation level for individual transaction.
+        # Here, we open a connection with SERIALIZABLE isolation level.
+        # it will be released when session.commit is called.
+        session = self.SessionMaker()
+        session.connection(execution_options={"isolation_level": "SERIALIZABLE"})
+
+        run = self._get_run(run_uuid=run_id, session=session)
+        self._check_run_is_active(run)
+
+        try:
+            current_tags = (
+                session.query(SqlTag)
+                .filter(SqlTag.run_uuid == run_id, SqlTag.key.in_([t.key for t in tags]))
+                .with_for_update()
+                .all()
+            )
+            current_tags = {t.key: t for t in current_tags}
+
+            new_tag_dict = {}
+            for tag in tags:
+                _validate_tag(tag.key, tag.value)
+                current_tag = current_tags.get(tag.key)
+                new_tag = new_tag_dict.get(tag.key)
+
+                # update the SqlTag if it is already present in DB
+                if current_tag:
+                    current_tag.value = tag.value
+                    continue
+
+                # if the SqlTag instance is recently created,
+                # update with incoming
+                if new_tag:
+                    new_tag.value = tag.value
+                # otherwise, put it into the dict
+                else:
+                    new_tag = SqlTag(run_uuid=run_id, key=tag.key, value=tag.value)
+
+                new_tag_dict[tag.key] = new_tag
+
+            # finally, save new entries to DB.
+            self._save_to_db(session=session, objs=list(new_tag_dict.values()))
+            session.commit()
+        except sqlalchemy.exc.IntegrityError:
+            session.rollback()
+            # two concurrent operations may try to attempt to insert tags.
+            # apply retry here.
+            max_retries = 3
+            attempt = (attempt or 0) + 1
+            if attempt > max_retries:
+                raise MlflowException(
+                    "Failed to set tags with given keys={} within {} retries".format(
+                        [t.key for t in tags], max_retries
+                    )
+                )
+            sleep_duration = 0.1 * ((2 ** attempt) - 1)
```

Comment:
I think we should increase the sleep duration slightly.

```suggestion
            sleep_duration = ((2 ** attempt) - 1)
```

Reply:
Good idea, done